src/Relations.dfy

```diff
@@ -49,6 +49,27 @@ module {:options "-functionSyntax:4"} Relations {
     forall i, j | 0 <= i < j < |a| :: lessThan(a[i], a[j])
   }
 
+  /* An element in an ordered set is called minimal, if it is less than every element of the set. */
+  ghost predicate IsMinimum<T>(R: (T, T) -> bool, m: T, s: set<T>) {
+    m in s && forall y: T | y in s :: R(m, y)
+  }
+
+  /* Any totally ordered set contains a unique minimal element. */
+  lemma LemmaUniqueMinimum<T(!new)>(R: (T, T) -> bool, s: set<T>) returns (m: T)
+    requires |s| > 0 && TotalOrdering(R)
+    ensures IsMinimum(R, m, s) && (forall n: T :: IsMinimum(R, n, s) ==> m == n)
+  {
+    var x :| x in s;
```

Comment:
I wonder if we should try to type variables explicitly? Is there some guideline on whether or not we want this for the library?

Reply:
The style guide http://dafny.org/dafny/StyleGuide/Style-Guide.html#type-declaration contains the guidance "If the type can be inferred by Dafny, leave it out, unless you think it provides useful documentation in the program.".